### PR TITLE
Allow chart labels to be provided together datasets

### LINKED ISSUE #2108: [Enhancement] Allow chart data to be provided as Object[]
- [X] I have written a [good issue](https://github.com/kirbydesign/designsystem/wiki/The-Good%3A-Issue) 

## Describe the enhancement
It's possible to provide data to the chart component as a [`Object[]`](https://www.chartjs.org/docs/master/general/data-structures.html#object) - allowing labels to be provided together with data. 
```typescript
  data: ChartDataset[] = [
    {
      data: [
        { x: 'a', y: 100 },
        { x: 'b', y: 1000 },
        { x: 'c', y: 120 },
      ],
    },
    {
      data: [
        { x: 'a', y: 100 },
        { x: 'b', y: 1000 },
        { x: 'c', y: 120 },
      ],
    },
  ];
```

We were unaware of this when developing the chart component. It therefore gives us two issues: 
1. Using the type `ChartDataset[]` to annotate the data, results in the following typescript error: `[tsserver 2322] [E] Type 'string' is not assignable to type 'number'.`

2. Blank labels are generated causing the chart to only take up half of the space: 
Screenshot 2022-03-15 at 12 50 46

We should fix that.

## Are there any additional context?
Problem 2 is related to the logic in #2085; therefore the simplification should be carried out as part of this issue, such that both #2085 and this is solved by the resulting PR.

### PR DESCRIPTION
**Pull request by MadsBuchmann**
## Which issue does this PR close?

This PR closes #2108

This PR also solves a bug originating from #kirby-guild which resulted in some stock charts displaying with 'invalid date' as labels. 

## What is the new behavior?
The logic has been simplified to no longer care about type and it has been extended to allow datasets with labels.

## Does this PR introduce a breaking change?

- [ ] Yes
- [X] No

## Checklist:

The following tasks should be carried out in sequence in order to follow [the process of contributing](https://github.com/kirbydesign/designsystem/blob/stable/.github/CONTRIBUTING.md/#the-process-of-contributing) correctly.

### Reminders
- [ ] Make sure you have implemented tests following the guidelines in: "[The good: Test](https://github.com/kirbydesign/designsystem/wiki/The-Good%3A-Test)".
- [ ] ~~Make sure you have updated the cookbook with examples and showcases (for bug fixes, enhancements & new components).~~

### Review  
- [ ] Do a [self-review](https://github.com/kirbydesign/designsystem/wiki/The-Good%3A-Self-review).
- [ ] Request that the changes are code-reviewed 
- [ ] ~~Request that the changes are [UX reviewed](https://github.com/kirbydesign/designsystem/blob/stable/.github/CONTRIBUTING.md/#ux-review) (only necessary if your PR introduces visual changes)~~

When the pull request has been approved it will be automatically merged to `stable` via [automerge](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/automatically-merging-a-pull-request).

### CODE REVIEW

File: libs/designsystem/src/lib/components/chart/chart-js/chart-js.service.ts [hunk 1]
```diff
@@ -271,6 +277,15 @@ export class ChartJSService {
     );
   }
 
+  private datasetsHasLabels(datasets: ChartDataset[], indexAxis = 'x') {
```
Comment (MadsBuchmann) [inline libs/designsystem/src/lib/components/chart/chart-js/chart-js.service.ts:280]:
```suggestion
  private datasetsContainsLabels(datasets: ChartDataset[], indexAxis = 'x') {
```

File: libs/designsystem/src/lib/components/chart/chart-js/chart-js.service.ts [hunk 2]
```diff
@@ -279,23 +294,22 @@ export class ChartJSService {
...
     const typeConfig = this.chartConfigService.getTypeConfig(type);
 
+    /* 
+       TODO cases: 
+       1. Labels are given apply them - Should solve Troels lenda timescale issue & Jonas Rodells issue [X]
+       2. No labels are given 
+        - if dataset has labels use those - should solve Jørgens way of passing in data []
+        - else if stock use default stock labels - Troels Lenda [X]
+        - else generate blank labels. Can we just... Not provide them? - solves case where no labels are provided.[X]
+    */
```
Comment (MadsBuchmann) [inline libs/designsystem/src/lib/components/chart/chart-js/chart-js.service.ts:304]:
```suggestion
```

File: libs/designsystem/src/lib/components/chart/chart-js/chart-js.service.ts [hunk 3]
```diff
@@ -271,6 +271,15 @@ export class ChartJSService {
     );
   }
 
+  private datasetsHaveLabels(datasets: ChartDataset[], indexAxis = 'x') {
+    if (isNumberArray(datasets)) return false;
+
+    return datasets.some(({ data }) => {
+      if (data === undefined) return false;
+      return data.some((datapoint) => typeof datapoint[indexAxis] === 'string');
+    });
```
Comment (mictro) [inline libs/designsystem/src/lib/components/chart/chart-js/chart-js.service.ts:280]: return datasets.some(
      ({ data }) =>
        data !== undefined &&
        data.some(
          (datapoint) => typeof datapoint === 'object' && typeof datapoint[indexAxis] === 'string'
        )
    );

File: libs/designsystem/src/lib/components/chart/chart-js/chart-js.service.ts [hunk 4]
```diff
@@ -271,6 +271,30 @@ export class ChartJSService {
     );
   }
 
+  private getLabelsToApply(args: {
+    datasets: ChartDataset[];
+    type: ChartType;
+    indexAxis: string;
+    labels?: unknown[];
+  }) {
```
Comment (mictro) [inline libs/designsystem/src/lib/components/chart/chart-js/chart-js.service.ts:279]: Add return type, is it (string[] | null) or ? - do the same to contributing methods where weak or missing.
Consider strengthning typinping of label parameter as well.

File: libs/designsystem/src/lib/components/chart/chart-js/chart-js.service.spec.ts [hunk 5]
```diff
@@ -170,18 +175,106 @@ describe('ChartJSService', () => {
...
+          expect(data.length).toBeGreaterThan(0);
+          expect(data.every(({ x }) => typeof x === 'string')).toBeTrue();
+
+          chartJSService.renderChart({
+            targetElement: canvasElement,
+            type: 'column',
+            data: dataset,
+          });
+
+          /* This assertion relies on the assumption that:
```
Comment (mictro) [inline libs/designsystem/src/lib/components/chart/chart-js/chart-js.service.spec.ts:267]: Nice with an explanation for the less obvious stuff.
Consider if the actual implementation also might benefit from an explanation regarding which labels are to be handled and how they are processed - assumtions etc.
It's very difficult to read from the code alone and weak or missing types doesn't help either.

**Review by mictro — CHANGES_REQUESTED** [review]
Excellent test coverage.
Added a few comments aiming at improving  the understanding of the intention behind the somewhat complex label handling in the component.

File: libs/designsystem/src/lib/components/chart/chart-js/chart-js.service.spec.ts [hunk 6]
```diff
@@ -170,18 +175,106 @@ describe('ChartJSService', () => {
...
+          1. if chart.data.labels is empty &
+          2. datasets contains string along the index axis 
+          : then the strings in datasets will be used as labels. 
+
+          This assumption comes from the internal logic of chartjs. 
+          */
+          const chartData = chartJSService['chart'].data;
```
Comment (mictro) [inline libs/designsystem/src/lib/components/chart/chart-js/chart-js.service.spec.ts:274]: ice with an explanation for the less obvious stuff. Consider if the actual implementation also might benefit from an explanation regarding which labels are to be handled and how they are processed - assumtions etc. It's very difficult to read from the code alone and weak or missing types doesn't help either.

File: libs/designsystem/src/lib/components/chart/chart-js/chart-js.service.spec.ts (hunk 5 above)
Comment (mictro) [inline libs/designsystem/src/lib/components/chart/chart-js/chart-js.service.spec.ts:267]: ice with an explanation for the less obvious stuff. Consider if the actual implementation also might benefit from an explanation regarding which labels are to be handled and how they are processed - assumtions etc. It's very difficult to read from the code alone and weak or missing types doesn't help either.

File: libs/designsystem/src/lib/components/chart/chart-js/chart-js.service.spec.ts [hunk 7]
```diff
@@ -170,18 +175,106 @@ describe('ChartJSService', () => {
...
+          expect(chartJSService['chart'].options.indexAxis).toEqual('x');
```
Comment (mictro) [inline libs/designsystem/src/lib/components/chart/chart-js/chart-js.service.spec.ts:275]: ice with an explanation for the less obvious stuff. Consider if the actual implementation also might benefit from an explanation regarding which labels are to be handled and how they are processed - assumtions etc. It's very difficult to read from the code alone and weak or missing types doesn't help either.

File: libs/designsystem/src/lib/components/chart/chart-js/chart-js.service.spec.ts [hunk 8]
```diff
@@ -170,18 +175,106 @@ describe('ChartJSService', () => {
...
+          chartJSService.renderChart({
+            targetElement: canvasElement,
+            type: 'column',
+            data: dataset,
+          });
+
+          /* This assertion relies on the assumption that:
+          1. if chart.data.labels is empty &
+          2. datasets contains string along the index axis 
+          : then the strings in datasets will be used as labels.
```
Comment (mictro) [inline libs/designsystem/src/lib/components/chart/chart-js/chart-js.service.spec.ts:270]: Nice with an explanation for the less obvious stuff. Consider if the actual implementation also might benefit from an explanation regarding which labels are to be handled and how they are processed - assumtions etc. It's very difficult to read from the code alone and weak or missing types doesn't help either.

File: libs/designsystem/src/lib/components/chart/chart-js/chart-js.service.ts (hunk 4 above)
Comment (MadsBuchmann) [reply libs/designsystem/src/lib/components/chart/chart-js/chart-js.service.ts:279]: I've added the following type to `chart.types.ts`: 
```typescript
export type ChartLabel = string | string[]; // String[] allows for multi-line labels
```

The `unknown[]` type has now been replaced with `ChartLabel[]`. This should strengthen the typing of the parameter.  

The `ChartLabel` type has also been added to the public API of Kirby.

File: libs/designsystem/src/lib/components/chart/chart-js/chart-js.service.spec.ts (hunk 8 above)
Comment (MadsBuchmann) [reply libs/designsystem/src/lib/components/chart/chart-js/chart-js.service.spec.ts:270]: I've attempted to make the label logic more readable and i've added comments. I hope it's more understandable now :D 
```typescript
  private getLabelsToApply(args: {
    datasets: ChartDataset[];
    type: ChartType;
    indexAxis: string;
    labels?: ChartLabel[];
  }) {
    const { datasets, labels, type, indexAxis } = args;

    const datasetHasLabels = ({ data }: ChartDataset) =>
      !!data?.some(
        (datapoint) => typeof datapoint === 'object' && typeof datapoint[indexAxis] === 'string'
      );

    /*
       Labels can be provided by the user two ways:
       1. As a seperate ChartLabel[] via the 'labels' input prop - this has highest priority
       2. Together with the dataset via the 'data' input prop - here each datapoint contains 
       a label for the indexAxis. 
       For example: { x: 'label1', y: 1} in the case where the index axis is 'x'. 

       If no labels are provided then default labels are used. 
    */
    const labelsAreGivenAsSeperateArray = labels !== undefined;
    const labelsAreGivenTogetherWithDataset = datasets.some(datasetHasLabels);

    if (labelsAreGivenAsSeperateArray) {
      return labels;
    } else if (labelsAreGivenTogetherWithDataset) {
      return null;
    } else {
      /* 
        Chart.js requires labels along the index axis to render anything therefore
        all other types than stock uses empty labels as default. The stock type 
        displays day & month as default. 
      */
      return type === 'stock'
        ? this.getDefaultStockLabels(datasets, this.locale)
        : this.createBlankLabels(datasets);
    }
  }
```

File: libs/designsystem/src/lib/components/chart/chart-js/chart-js.service.ts (hunk 4 above)
Comment (mictro) [reply libs/designsystem/src/lib/components/chart/chart-js/chart-js.service.ts:282]: Sweet :=)

**Review by mictro — APPROVED** [review]

**Review by mictro — DISMISSED** [review]

**Review by mictro — APPROVED** [review]

**Pull request merged**